src/panels/media-browser/ha-panel-media-browser.ts

```diff
@@ -0,0 +1,184 @@
+              .narrow=${this.narrow}
+            ></ha-menu-button>
+            <div main-title>
+              ${title || ""}${this.hass.localize(
+                "ui.components.media-browser.media-player-browser"
+              )}
+            </div>
+            <mwc-icon-button
+              .label=${this.hass.localize(
+                "ui.components.media-browser.choose-source"
+              )}
+              @click=${this._showMediaSouceDialog}
+            >
+              <ha-svg-icon .path=${mdiLoginVariant}></ha-svg-icon>
+            </mwc-icon-button>
+          </app-toolbar>
+        </app-header>
+        <div class="content">
+          ${this._entityId
+            ? html`
+                <ha-media-player-browse
+                  .hass=${this.hass}
+                  .entityId=${this._entityId}
+                  @media-picked=${this._mediaPicked}
+                ></ha-media-player-browse>
+              `
+            : html`
+                <div class="no-source">
+                  <h2>
+                    ${this.hass.localize(
+                      "ui.components.media-browser.choose_media_source"
+                    )}
+                  </h2>
+                  <mwc-button raised @click=${this._showMediaSouceDialog}
+                    >${this.hass.localize(
+                      "ui.components.media-browser.choose-source"
+                    )}</mwc-button
+                  >
+                </div>
+              `}
+        </div>
+      </ha-app-layout>
+    `;
+  }
+
+  private _showMediaSouceDialog(): void {
+    showSelectMediaSourceDialog(this, {
+      mediaSources: this._browseMediaSources(this.hass),
+      sourceSelectedCallback: (entityId) => {
+        this._entityId = entityId;
+        this.requestUpdate();
+      },
+    });
+  }
+
+  private async _mediaPicked(
+    ev: HASSDomEvent<MediaPickedEvent>
+  ): Promise<void> {
+    const item = ev.detail.item;
+    if (this._entityId === BROWSER_SOURCE) {
+      const resolvedUrl: any = await this.hass.callWS({
+        type: "media_source/resolve_media",
+        media_content_id: item.media_content_id,
+      });
+
+      showMediaPlayerDialog(this, {
+        sourceUrl: resolvedUrl.url,
+        sourceType: resolvedUrl.mime_type,
+        title: item.title,
+      });
+      return;
+    }
+
+    this.hass!.callService("media_player", "play_media", {
+      entity_id: this._entityId,
+      media_content_id: item.media_content_id,
+      media_content_type: item.media_content_type,
+    });
+  }
+
+  private _browseMediaSources = memoize((hass: HomeAssistant) => {
+    if (!hass) {
+      return [];
+    }
+
+    let entityIds = Object.keys(hass.states);
+
+    entityIds = entityIds.filter((eid) =>
+      ["media_player"].includes(computeDomain(eid))
+    );
+
+    entityIds = entityIds.filter((eid) =>
+      supportsFeature(hass.states[eid], SUPPORT_BROWSE_MEDIA)
+    );
+
+    return entityIds.map((key) => hass.states[key]);
```

Comment:
And combine it with this map too, not need to walk it 3 times, we can do it in 1.

```ts
const entities = [];
Object.values(hass.states).forEach(entity => {
  if (computeStateDomain(entity) === "media_player" && supportsFeature(entity, SUPPORT_BROWSE_MEDIA)) {
    entities.push(entity);
  }
});
return entities;